raft_node.py

```diff
@@ -0,0 +1,595 @@
+    - AppendEntries: a leader will send this to any other node where there are
+      entries in the leader's log that have not been replicated on that node.
+      The leader will also regularly send empty AppendEntriesRequests to all
+      other nodes; these function as heartbeats. Recipients will add the new
+      log entries to their own logs, so long as the leader's log is consistent
+      with their own. (In the event that it is not, the leader will attempt to
+      find a point in the past at which the recipient's log is in agreement and
+      then start replication from there, potentially overwriting some entries
+      that were not yet committed.)
+    - VoteRequest: when a node becomes a candidate, it will start an election
+      by incrementing its term and sending VoteRequests to all other nodes. A
+      recipient will vote for any candidate whose term is greater than its own,
+      so long as they have not already voted for a different candidate for the
+      same term.
+    - GetValue: this will retrieve a mapping from the current state. In this
+      implementation, any node is allowed to respond to a GetValue request. If
+      the responder is not the leader, this means the client may receive a
+      stale value (but it will never see a value that has not been committed).
+    - ProposeMapping: the leader will add the mapping to its own log, attempt
+      to replicate it on a majority of nodes in the cluster, and on success
+      respond to the client. If a non-leader receives this RPC, it will respond
+      with the identity of the current leader so the client may retry.
+"""
+
+
+import aiofiles
+import asyncio
+import json
+import os
+import random
+import sys
+from asyncio.exceptions import CancelledError, TimeoutError
+from collections.abc import Coroutine, Iterable
+from dataclasses import dataclass, field
+from datetime import timedelta
+from typing import Any, Callable, Self, TypeVar
+
+from grpc.aio._call import AioRpcError
+
+import cluster_server
+import raft_server
+from log import Log
+from node_state import CandidateState, FollowerState, LeaderState, NodeState
+from raft_client import Client
+from rpc.raft_pb2 import (
+    AppendEntriesRequest,
+    AppendEntriesResponse,
+    GetValueResponse,
+    LogEntry,
+    Key,
+    Mapping,
+    ProposeMappingResponse,
+    RequestVoteResponse,
+    VoteRequest,
+)
+from task_scheduler import run_in_background, run_every
+
+
+@dataclass(slots=True, frozen=True)
+class Config:
+    raft_server_port: int
+    cluster_server_port: int
+    all_raft_server_ports: list[int]
+    heartbeat_timeout_ms: int
+    min_election_timeout_ms: int
+    max_election_timeout_ms: int
+    append_entries_timeout_ms: int
+    log_file: str
+    persistent_state_file: str
+
+
+Req = TypeVar("Req", AppendEntriesRequest, VoteRequest)
+Nd = TypeVar("Nd", bound="Node")
+Rsp = TypeVar("Rsp")
+
+
+def check_term(f: Callable[[Nd, Req], Rsp]) -> Callable[[Nd, Req], Rsp]:
```

Comment:
I'd claim that it's an antipattern to put this kind of nontrivial logic into a decorator, although I don't think that I'm obviously right here.

I'd prefer to just start each of the functions below with:
```
self.update_term(rpc_request.term, rpc_request.requester_id)
self.reset_heartbeat_timer(rpc_request.request_id)
```

I also don't really understand why we only update heartbeats if the term is current - it doesn't seem necessary to me, but I could be wrong.

Reply:
On the other hand, I _would_ support a decorator for @leader_only, that abstracts out most of the _propose_mapping_handler, and passes the leader state as an extra argument to [f]. Maybe my actual gripe with this decorator is that it does a bunch of side-effect things, and then just passes the original arguments down to [f] unchanged.